test/index.test.js

```diff
@@ -70,160 +73,267 @@ test(file + " compile .txt template", function (t) {
   t.end()
 });
 
-test(file + " Force Fail in Email for a single email", function (t) {
-  var person = {
-    "name": "Bounce",
-    "email": "invalid.email.address",
-    "subject": "Welcome to DWYL :)"
+
+//// EMAIL SPECIFIC TESTS
+
+// For easier access to sending functions
+// We reassign them in email_test_startup, run before each sending test
+var email, sendMany;
+
+/**
+ * Sets state for each test on startup, using a specific service
+ * when requested (and requested service is valid and configured).
+ *
+ * @param {String} [current_service_name] - name of a sending service
+ * @returns {undefined}
+ */
+function email_test_startup (current_service_name) {
+
+  var SERVICE_TESTING = {
+    mailgun: {
+      errorKey: "statusCode",
+      successIdKey: "id",
+      // Mailgun doesn't offer a test mailbox, so you need to authorize a recipient in your dashboard
+      successSimulator: process.env.MAILGUN_AUTHORIZED_RECIPIENT,
+    },
+    ses: {
+      errorKey: "statusCode",
+      successIdKey: "MessageId",
+      successSimulator: "success@simulator.amazonses.com"
+    }
   };
-  email('hello', person, function (err) {
-    t.equal(err.statusCode, 400, "Invalid Mandrill Key");
-    t.end();
+
+  var specific_service_config = SERVICE_TESTING[current_service_name];
+  process.env.SENDEMAIL_SERVICE = current_service_name;
+  process.env.SUCCESS_SIMULATOR = specific_service_config.successSimulator;
+  process.env.SUCCESS_ID_KEY = specific_service_config.successIdKey;
+  process.env.ERROR_KEY = specific_service_config.errorKey;
+
+  // We decache the library's index file
+  // so we can re-require it, thereby rerunning
+  // service.initialize based on our changed env settings
+  decache('../lib/index.js');
+  sendemail = require('../lib/index.js');
+  email = sendemail.email;
+  sendMany = sendemail.sendMany;
+}
+
+/**
+  In order to ensure we rerun service init, we need to decache the main lib
+  What happens to the the .email  and sendMany refs?
+**/
+
+/**
+ * Removes all configuration set at test runtime, ensuring tests
+ * are totally independent, configuration-wise.
+ *
+ * @returns {undefined}
+ */
+function email_test_teardown () {
+
+  delete process.env.SENDEMAIL_SERVICE;
+  delete process.env.SUCCESS_SIMULATOR;
+  delete process.env.SUCCESS_ID_KEY;
+  delete process.env.ERROR_KEY;
+}
+
+service.service_configs
+  .map(function(conf) {
+    return conf.name;
+  })
+  .forEach(function (specific_service) {
+  // Annotates tests with current sending service
+  var service_test_description = file + " — " + (specific_service || "Default Sender") + ":";
+
+  test(service_test_description + " Force Fail in Email for a single email", function (t) {
+    email_test_startup(specific_service);
```

Comment:
I see why you've done this `startup` and `teardown`, but it would be nice if we could make `email` a "purer" function which doesn't read env vars and simplify this test

Reply:
Hmmm...gotcha, yea, I'd have to think on that one. But thanks for the feedback, though. Agree that'd be a good move; I definitely made the tests a good bit more.... busy :)